File: 7_Kyu_Kata/make-class/Python/make_class_test.py
# [object Object] - Make Class  [ ID: 5d774cfde98179002a7cb3c8  (make-class) ]
# URL: https://www.codewars.com/kata/5d774cfde98179002a7cb3c8
# Category: ALGORITHMS  |  Tags: OBJECT-ORIENTED PROGRAMMING | ALGORITHMS
# ******************************************************************************
@test.describe("Fixed tests")
def fixed_tests():
    @test.it("Tests")
    def it_1():
        class Animal:
            def __init__(self, name, species, age, health, weight, color):
                self.name = name
                self.species = species
                self.age = age
                self.health = health
                self.weight = weight
                self.color = color
        
        Animel = make_class("name", "species", "age", "health", "weight", "color")
        
        dog1 = Animal("Bob", "Dog", 5, "good", "50lb", "brown")
        dog2 = Animel("Bob", "Dog", 5, "good", "50lb", "brown")
        
        test.assert_equals(dog1.name, dog2.name)
        test.assert_equals(dog1.species, dog2.species)
        test.assert_equals(dog1.age, dog2.age)
        test.assert_equals(dog1.health, dog2.health)
        test.assert_equals(dog1.weight, dog2.weight)
        test.assert_equals(dog1.color, dog2.color)

@test.describe("Random tests")
def random_tests():
    from random import randint as R, choices as C
    from string import ascii_lowercase as AL

    def solution(*args):
        class F:
            def __init__(self, *brgs):
                self.__dict__ = {x: y for x, y in zip(args, brgs)}
        return F

    @test.it("Tests")
    def it_1():
        for _ in range(10):
            n = R(1, 8)
            s = set()
            while len(s) < n:
                s.add("".join(C(AL, k=R(5, 10))))
            s = [*s]
            result, user = solution(*s), make_class(*s)
            for _ in range(10):
                a = [R(1, 1000) if R(0, 1) else "".join(C(AL, k=R(3, 5))) for _ in range(n)]
                test.assert_equals(user(*a).__dict__, result(*a).__dict__)

File: kata-7-kyu/make-class/javascript/makeClass.Test.js
/* eslint-disable no-undef */
//+ ====================================================================================================================
//+
//+ 7 kyu - Make Class  [ ID: 5d774cfde98179002a7cb3c8 ] (make-class)
//+ URL: https://www.codewars.com/kata/5d774cfde98179002a7cb3c8
//+ Category: ALGORITHMS  |  Tags: OBJECT-ORIENTED PROGRAMMING | ALGORITHMS
//+
//+ ====================================================================================================================

import { describe } from "node:test"
import { assert } from "vitest"
const { makeClass } = require("./makeClass")

function solution(...keys) {
  return function (...values) {
    keys.forEach((key, i) => {
      this[key] = values[i]
    })
  }
}

function Animal(name, species, age, health, weight, color) {
  this.name = name
  this.species = species
  this.age = age
  this.health = health
  this.weight = weight
  this.color = color
}
const Animel = makeClass("name", "species", "age", "health", "weight", "color")

describe("Basic Tests", function () {
  const dog1 = new Animal("Bob", "Dog", "5", "good", "50lb", "brown")
  const dog2 = new Animel("Bob", "Dog", "5", "good", "50lb", "brown")
  it("check for object similarity", function () {
    assert.strictEqual(Object.keys(dog1).join`,`, Object.keys(dog2).join`,`)
    assert.strictEqual(Object.values(dog1).join`,`, Object.values(dog2).join`,`)
  })
  it("anti-cheat", function () {
    assert.strictEqual(
      dog2 instanceof Animel,
      true,
      '"instances" of your class need to be instances of your class.',
    )
    assert.strictEqual(dog2.constructor === Animel, true, "what are you doing")
  })
})

describe("Random Tests", function () {
  function Obj(x, y, z) {
    this.x = x
    this.y = y
    this.z = z
  }
  const ObjClass = makeClass("x", "y", "z")

  it("Some random Tests", function () {
    for (let i = 0; i < 50; i++) {
      const [a, b] = [Math.random() * 1000, Math.random() * 1000]
      const o = new Obj(a, b, (a, b) => a + b)
      const p = new ObjClass(a, b, (a, b) => a + b)
      assert.strictEqual(
        JSON.stringify(o),
        JSON.stringify(p),
        "JSON representation of object instances must be the same",
      )
      assert.strictEqual(p.z(p.x, p.y), o.z(o.x, o.y))
    }
  })

  it("Some more random Tests", function () {
    for (let i = 0; i < 20; i++) {
      const parameters = Array.from(
        new Set([...Array(26)].map(() => String.fromCharCode((Math.random() * 25 + 65) | 0))),
      )

      function MyClass(...parameters) {
        parameters.reduce((a, p) => a + p, "")
      }

      // const MyClass = solution(...parameters)
      const YourClass = makeClass(...parameters)

      assert.strictEqual(JSON.stringify(MyClass), JSON.stringify(YourClass))
      for (let i = 0; i < 20; i++) {
        const randomArguments = [...("" + ((Math.random() * Number.MAX_SAFE_INTEGER) | 0))]
        const myInstance = new MyClass(...randomArguments)
        const yourInstance = new YourClass(...randomArguments)
        assert.strictEqual(
          JSON.stringify(myInstance),
          JSON.stringify(yourInstance),
          "JSON representation of object instances must be the same",
        )
      }
    }
  })
})
